pgrx-pg-sys/src/submodules/elog.rs

```diff
@@ -137,14 +327,109 @@ macro_rules! debug4 {
+///     hint = "{} helpful", "very";
+/// }
+/// ```
 #[macro_export]
 macro_rules! debug3 {
+    (
+        $msg:literal $(, $msg_args:expr)*;
+        $(detail = $detail:literal $(, $detail_args:expr)*)?;
+    ) => (
+        {
+            extern crate core;
+            extern crate alloc;
+            $crate::ereport!(
+                $crate::elog::PgLogLevel::DEBUG3,
+                $crate::errcodes::PgSqlErrorCode::ERRCODE_SUCCESSFUL_COMPLETION,
+                alloc::format!("{}", core::format_args!($msg $(, $msg_args)*));
+                $(detail = alloc::format!("{}", core::format_args!($detail $(, $detail_args)*));)?
+            );
+        }
+    );
+    (
+        $msg:literal $(, $msg_args:expr)*;
+        $(hint = $hint:literal $(, $hint_args:expr)*)?;
+    ) => (
+        {
+            extern crate core;
+            extern crate alloc;
+            $crate::ereport!(
+                $crate::elog::PgLogLevel::DEBUG3,
+                $crate::errcodes::PgSqlErrorCode::ERRCODE_SUCCESSFUL_COMPLETION,
+                alloc::format!("{}", core::format_args!($msg $(, $msg_args)*));
+                $(hint = alloc::format!("{}", core::format_args!($hint $(, $hint_args)*));)?
+            );
+        }
+    );
+    (
+        $msg:literal $(, $msg_args:expr)*;
+        $(hint = $hint:literal $(, $hint_args:expr)*)?;
+        $(detail = $detail:literal $(, $detail_args:expr)*)?;
+    ) => (
+        {
+            extern crate core;
+            extern crate alloc;
+            $crate::ereport!(
+                $crate::elog::PgLogLevel::DEBUG3,
+                $crate::errcodes::PgSqlErrorCode::ERRCODE_SUCCESSFUL_COMPLETION,
+                alloc::format!("{}", core::format_args!($msg $(, $msg_args)*));
+                $(detail = alloc::format!("{}", core::format_args!($detail $(, $detail_args)*));)?
+                $(hint = alloc::format!("{}", core::format_args!($hint $(, $hint_args)*));)?
+            );
+        }
+    );
+    (
+        $msg:literal $(, $msg_args:expr)*;
+        $(detail = $detail:literal $(, $detail_args:expr)*)?;
+        $(hint = $hint:literal $(, $hint_args:expr)*)?;
+    ) => (
+        {
+            extern crate core;
+            extern crate alloc;
+            $crate::ereport!(
+                $crate::elog::PgLogLevel::DEBUG3,
+                $crate::errcodes::PgSqlErrorCode::ERRCODE_SUCCESSFUL_COMPLETION,
+                alloc::format!("{}", core::format_args!($msg $(, $msg_args)*));
+                $(detail = alloc::format!("{}", core::format_args!($detail $(, $detail_args)*));)?
+                $(hint = alloc::format!("{}", core::format_args!($hint $(, $hint_args)*));)?
+            );
+        }
+    );
     ($($arg:tt)*) => (
         {
             extern crate alloc;
             $crate::ereport!($crate::elog::PgLogLevel::DEBUG3, $crate::errcodes::PgSqlErrorCode::ERRCODE_SUCCESSFUL_COMPLETION, alloc::format!($($arg)*).as_str());
         }
-    )
+    );
```

Comment:
This code has a rather lot of duplication now. That might have been tolerable when there was less code overall, but I should note that Rust allows you to define a macro that itself mostly serves to define new macros when invoked. It's an extremely "now you have two problems" sort of decision to go down that road, but this is at the point where it seems justified, doesn't it?

Reply:
1. I agree this is too much duplication. I wanted to avoid introducing a new public macro but if you think it is justified then I can refactor to keep things DRY. 
2. When you say "define a macro that itself mostly serves to define new macros", do you mean a private `macro by example` that defines another macro that is public? Or a functional proc macro that expands to a public macro? I was under the impression that it wasn't possible or at least wasn't practical to implement higher-order macros with only `macro by example`.

Reply:
A new public macro is not required.

It's possible and in my opinion practical, but it can be a bit of a trick, admittedly.

Reply:
If you hit me up on [Discord](https://discord.gg/9tAcg7w7), I can help walk you through the tricks more exactly (it's smoothest), but it will be best if you review [The Little Book of Rust Macros](https://veykril.github.io/tlborm/introduction.html) if you haven't, including the [callback](https://veykril.github.io/tlborm/decl-macros/patterns/callbacks.html) pattern. It may help explain some of the limitations of macros that you encountered.

But if the cleanest solution is to add a new public macro or something, that's actually fine? In particular, I have for some time now thought all the `debugN!` macros should just be a single:
```rust
#[macro_export]
macro_rules! debug {
    (1, $($tok:tt)) => {};
    (2, $($tok:tt)) => {};
    (3, $($tok:tt)) => {};
    (4, $($tok:tt)) => {};
    (5, $($tok:tt) => {};
```

We can match on literals to some extent, so we probably should.

Reply:
I refactored things to get things more DRY. When I run tests locally _(including ui tests)_ everything passes but the GH checks have an issue comparing one of the ui snapshots. Can you help me understand what the issue is?

Also any feedback you have on this updated implementation, I would be grateful to hear it. I can hop on a call via Discord tomorrow if it would be easier to discuss over VoIP.